README.md
# rb-parsetree19

A C extension to support ParseTree in Ruby 1.9

## Requirements

A patched version of Ruby 1.9.2 is needed. In fact the same patched version
as needed by rb-threadframe (and thus trepanning).

Usage:

    require 'parse_tree'
    processor = ParseTree19.new(false)
    processor.parse_tree_for_str('x=1; y=2', '(string'), 1)
      => [:scope, [:block, [:lasgn, :x, [:lit, 1]], [:block, [:lasgn, :y, [:lit, 2]]]]]
    def five; 5 end
    processor.parse_tree_for_method(method(:five), true)
      => [:scope, [:args, 0, [:args_aux]], [:lit, 5]]
    puts processor.parse_tree_for_method(method(:five, false)
      => nil
    ###########################################################
    ## Do NOT use this node dump for any purpose other than  ##
    ## debug and research.  Compatibility is not guaranteed. ##
    ###########################################################
    
    # @ NODE_SCOPE (line: 6)
    # +- nd_tbl: (empty)
    # +- nd_args:
    # |   @ NODE_ARGS (line: 6)
    # |   +- nd_frml: 0
    # |   +- nd_next:
    # |   |   @ NODE_ARGS_AUX (line: 6)
    # |   |   +- nd_rest: (null)
    # |   |   +- nd_body: (null)
    # |   |   +- nd_next:
    # |   |       (null node)
    # |   +- nd_opt:
    # |       (null node)
    # +- nd_body:
    #     @ NODE_LIT (line: 6)
    #     +- nd_lit: 5


NEWS
June 12, 2011 (0.3)
 Add NODE_BLOCK and NODE_BLOCK_ARG 
 Guard against dereferening a NULL iseq pointer
 Remove some gcc warnings

December 25, 2010 (0.2)
 First public release


test/unit/test_parse_tree_for_str.rb
require 'test/unit'
require_relative '../../ext/parse_tree'

class TestParseTree < Test::Unit::TestCase
  def setup
    @processor = ParseTree19.new(false)
  end

  def test_parse_tree_for_str
    # The order of the ops below is the order they appear in
    # add_to_parse_tree.inc/node.c
    # Note: if you ever wanted have a Ruby expression that causes
    # a given parse node to be created, see below.
    [
     # block:
     ['foo; bar',      [:scope, [:block, [:vcall, :foo], 
                                 [:block, [:vcall, :bar]]]]],

     # if:
     ['true if nil',   [:scope, [:if, [:nil], [:true]]]],
     ['if nil then true else false end', 
      [:scope, [:if, [:nil], [:true], [:false]]]],

     # Need case, when, opt_n

     ['5 while false', [:scope, [:while, [:false], [:lit, 5]]]],
     ['4 until true',  [:scope, [:until, [:true], [:lit, 4]]]],
     
     # iter:
     ['x { 5 }', 
      [:scope, [:iter, [:fcall, :x], [:scope, [:lit, 5]]]]],

     # Need for
       
     ['break',         [:scope, [:break]]],
     ['break 1',       [:scope, [:break, [:lit, 1]]]],
     ['next',          [:scope, [:next]]],
     ['next 1',        [:scope, [:next,  [:lit, 1]]]],
     ['return',        [:scope, [:return]]],
     ['return 4',      [:scope, [:return, [:lit, 4]]]],

     ['redo',          [:scope, [:redo]]],
     ['retry',         [:scope, [:retry]]],

     # begin
     ['begin; 1 ; end', 
      [:scope, [:block, [:begin], [:block, [:lit, 1]]]]],

     # rescue and resbody
     ['begin; 1 ; rescue; 2; else 3; end',
      [:scope, [:rescue, [:block, [:begin], [:block, [:lit, 1]]], 
                [:resbody, [:lit, 2]], [:lit, 3]]]],

     # ensure
     ['begin; 1 ; ensure; 2; end',
      [:scope, [:ensure, [:block, [:begin], [:block, [:lit, 1]]], 
                [:block, [:begin], [:block, [:lit, 2]]]]]],

     # and
     ['true and nil',  [:scope, [:and, [:true], [:nil]]]],
     # or
     ['true or false', [:scope, [:or,  [:true], [:false]]]],

     # lasgn
     ['x = 5',         [:scope, [:lasgn,  :x,   [:lit, 5]]]],
     # dasgn:
     ['x = nil; 1.times { x = 5 }',
      [:scope, [:block, [:lasgn, :x, [:nil]],
                [:block, [:iter, [:call, :times, [:lit, 1]], 
                          [:scope, [:dasgn, :x, [:lit, 5]]]]]]]],
     # dasgn_cur:
     ['1.times { x = 5 }',
      [:scope, [:iter, [:call, :times, [:lit, 1]], 
                [:scope, [:dasgn_curr, :x, [:lit, 5]]]]]],

     ['@x = 5',      [:scope, [:iasgn,  :@x,  [:lit, 5]]]],
     ['@@x = 5',     [:scope, [:cvasgn, :@@x, [:lit, 5]]]],

     ['$x = 5',      [:scope, [:gasgn,  :$x,  [:lit, 5]]]],

     ['x &&= 1',     [:scope, [:op_asgn_and,  
                               [:lvar, :x], [:lasgn, :x, [:lit, 1]]]]],
     ['x ||= 1',     [:scope, [:op_asgn_or,   
                               [:lvar, :x], [:lasgn, :x, [:lit, 1]]]]],

     # op_asgn1
     # 'ary[1] += 2'

     # call:
     ['x.foo',         [:scope, [:call, :foo, [:vcall, :x]]]],
       
     # For fcall, see iter above.

     ['foo',           [:scope, [:vcall, :foo]]],

     ['super',         [:scope, [:zsuper]]],

     # array
     ['[1]',           [:scope, [:array, 1, [:lit, 1]]]],

     # values array 1st item is not predictable.

     # zarray
     ['[]',            [:scope, [:zarray]]],

     # hash
     ['{}',            [:scope, [:hash]]],

     ['yield',         [:scope, [:yield]]],

     # For lvar, see x ||= 1

     # dvar: 
     ['1.times { x = 1; x }',
      [:scope, [:iter, [:call, :times, [:lit, 1]], 
                [:scope, [:block, [:dasgn_curr, :x, [:lit, 1]], 
                          [:block, [:dvar, :x]]]]]]],

     ['@x',            [:scope, [:ivar,  :@x]]],
     ['X',             [:scope, [:const, :X]]],
     ['@@X',           [:scope, [:cvar,  :@@X]]],
     ['$x',            [:scope, [:gvar,  :$x]]],

     ["5",             [:scope, [:lit, 5]]],
     ["'5'",           [:scope, [:str, '5']]],
     ['%x(5)',         [:scope, [:xstr, '5']]],
     ['"#{5}"',        [:scope, [:dstr, nil, [:evstr, [:lit, 5]]]]],
     ['`#{5}`',        [:scope, [:dxstr, nil, [:evstr, [:lit, 5]]]]],
     ['/#{5}/',        [:scope, [:dregx, '', [:evstr, [:lit, 5]]]]],
     ['/#{5}/o',       [:scope, [:dregx_once, '', [:evstr, [:lit, 5]]]]],
     [':"#{5}"',       [:scope, [:dsym, nil, [:evstr, [:lit, 5]]]]],

     ['alias a b',     [:scope, [:alias, [:lit, :a], [:lit, :b]]]],
     ['alias $a $b',   [:scope, [:valias, :$a, :$b]]],

     ['undef foo',     [:scope, [:undef, [:lit, :foo]]]],

     # dot2
     ['1..5.each{ 5 }',
      [:scope, [:dot2, [:lit, 1], 
                [:iter, [:call, :each, [:lit, 5]], [:scope, [:lit, 5]]]]]],
     # dot3
     ['1...5.each{ 5 }',
      [:scope, [:dot3, [:lit, 1], 
                [:iter, [:call, :each, [:lit, 5]], [:scope, [:lit, 5]]]]]],

     # Need flip2 and flip3

     ['self',          [:scope, [:self]]],
     ['nil',           [:scope, [:nil]]],
     ['true',          [:scope, [:true]]],
     ['false',         [:scope, [:false]]],

     ['M::Foo',        [:scope, [:colon2, :Foo, [:const, :M]]]],
     ['::Foo',         [:scope, [:colon3, :Foo]]],
     ['defined?(5)',   [:scope, [:defined, [:lit, 5]]]],
    ].each do |test_str, expected|
      actual   = @processor.parse_tree_for_str test_str, '(string)', 1
      assert_equal expected, actual
    end
  end
end


test/unit/test_parse_tree_for_meth.rb
require 'test/unit'
require_relative '../../ext/parse_tree'

def five; 5 end

class TestParseTree < Test::Unit::TestCase
  def setup
    @processor = ParseTree19.new(false)
  end

  def test_parse_tree_for_method

    def six
      x = 2
      y = 3
      return x * y
    end

    [
     [:five,      
      [:scope, [:args, 0, [:args_aux]], [:lit, 5]]],

     [:six, 
      [:scope, [:args, 0, [:args_aux]], 
       [:block, [:lasgn, :x, [:lit, 2]], 
        [:block, [:lasgn, :y, [:lit, 3]], 
         [:block, [:call, :*, [:lvar, :x], [:array, 1, [:lvar, :y]]]]]]]]
    ].each do |meth, expected|
      actual  = @processor.parse_tree_for_method(method(meth), true)
      assert_equal expected, actual
    end
  end
end


ext/parse_tree.c
/* What release we got? */
#define PARSETREE19_VERSION "0.3"  
#include "../include/vm_core_mini.h"   /* Pulls in ruby.h and node.h */
#include "../include/ruby19_externs.h" 

/* Turn a Ruby node string, e.g. "NODE_FALSE" into a symbol that ParseTree
   uses, e.g. :false. */
static VALUE 
pt_node_name(const char *name)
{
    if (0 == strncmp(name, "NODE_", strlen("NODE_"))) { 
	const char * name_sans_node = &name[strlen("NODE_")];
	return 
	    rb_funcall(
		rb_funcall(rb_str_new2(name_sans_node), rb_intern("downcase"), 
			   0),
		rb_intern("to_sym"), 0);
    } else {
	rb_bug("pt_node_name: node %s name should start with NODE_", name);
	return Qnil;
    }
}

static VALUE 
wrap_into_node(const char * name, VALUE val) 
{
    VALUE n = rb_ary_new();
    rb_ary_push(n, pt_node_name(name));
    if (val) rb_ary_push(n, val);
    return n;
}

#include "add_to_parse_tree.inc" 

static VALUE 
parse_tree_for_node(VALUE self, NODE *node, VALUE tree)
{
    VALUE result = rb_ary_new();

    if (Qfalse == tree)
	result = rb_parser_dump_tree(node, 0);
    else
	result = add_to_parse_tree(self, result, node, NULL);
    return (result);
}

static VALUE 
parse_tree_for_iseq_internal(VALUE self, rb_iseq_t *iseq, VALUE tree)
{
    VALUE result = rb_ary_new();

    if (iseq && iseq->tree_node) {
	if (Qfalse == tree)
	    result = rb_parser_dump_tree(iseq->tree_node, 0);
	else
	    result = add_to_parse_tree(self, result, iseq->tree_node, NULL);
	return (result);
    } else 
	return Qnil;
}

static VALUE 
parse_tree_for_iseq(VALUE self, VALUE iseqval, VALUE tree)
{
    rb_iseq_t *iseq;

    GetISeqPtr(iseqval, iseq);
    return parse_tree_for_iseq_internal(self, iseq, tree);
}

/* Defined in Ruby 1.9 proc.c */
extern rb_iseq_t *rb_method_get_iseq(VALUE method);

static VALUE 
parse_tree_for_method(VALUE self, VALUE method, VALUE tree)
{
    rb_iseq_t *iseq = rb_method_get_iseq(method);
    return parse_tree_for_iseq_internal(self, iseq, tree);
}

static VALUE 
parse_tree_common(VALUE self, VALUE source, VALUE filename, VALUE line,
                  VALUE tree)
{
    VALUE tmp;
    NODE *node = NULL;

#ifdef FIXED
    rb_thread_t *th = GET_THREAD();
    int critical;
#endif

    tmp = rb_check_string_type(filename);
    if (NIL_P(tmp)) {
	filename = rb_str_new2("(string)");
    }

    if (NIL_P(line)) {
	line = LONG2FIX(1);
    }

#ifdef FIXED
    ruby_nerrs = 0;
    critical = rb_thread_critical;
    rb_thread_critical = Qtrue;
    th->parse_in_eval++;
#endif

    StringValue(source);
    node = rb_compile_string(StringValuePtr(filename), source, NUM2INT(line));

#ifdef FIXED
    th->parse_in_eval--;
    rb_thread_critical = critical;
    if (ruby_nerrs > 0) {
	ruby_nerrs = 0;
	rb_exc_raise(th->errinfo);
    }
#endif

    return parse_tree_for_node(self, node, tree);
}

static VALUE 
parse_tree_for_str(VALUE self, VALUE source, VALUE filename, VALUE line) 
{
    return parse_tree_common(self, source, filename, line, Qtrue);
}

static VALUE 
parse_tree_dump(VALUE self, VALUE source, VALUE filename, VALUE line) 
{
    return parse_tree_common(self, source, filename, line, Qfalse);
}

void Init_parse_tree(void) 
{
    VALUE c = rb_define_class("ParseTree19", rb_cObject);
    rb_define_method(c, "parse_tree_for_iseq", 
		     (VALUE(*)(ANYARGS))parse_tree_for_iseq, 2);
    rb_define_method(c, "parse_tree_for_method", 
		     (VALUE(*)(ANYARGS))parse_tree_for_method, 2);
    rb_define_method(c, "parse_tree_for_str", 
		     (VALUE(*)(ANYARGS))parse_tree_for_str, 3);
    rb_define_method(c, "parse_tree_dump", 
		     (VALUE(*)(ANYARGS))parse_tree_dump, 3);
}
